### PR TITLE
Remote otadev

### PR DESCRIPTION
**Pull request by chezlyne**

### CODE REVIEW

File: lib/ZumoOta/src/App.cpp [hunk 1]
```diff
@@ -74,45 +83,66 @@ App::~App()
...
+
+        /* Set severity of logging system. */
+        Logging::getInstance().setLogLevel(CONFIG_LOG_SEVERITY);
+
+        LOG_DEBUG("LOGGER READY");
+        return true;
+    }
+    else 
+    {
+        LOG_ERROR("Fail to init Logging!");
```
Comment (nhjschulz) [inline lib/ZumoOta/src/App.cpp:100]: Does this have any value if the sink register failed in line 88 ? Guess it doesn't hurt either, but expect logging to not work. I would delete the else.

File: lib/ZumoOta/src/App.cpp [hunk 2]
```diff
@@ -74,45 +83,66 @@ App::~App()
...
     if (m_fileManager.init())
     {
-        Serial.println("LittleFS initialization successful");
+        LOG_DEBUG("LittleFS initialization successful");
         m_webServer.init();
     }
     else
     {
-        Serial.println("LittleFS initialization failed. The application will not start.");
+        LOG_FATAL("LittleFS initialization failed. The application will not start.");
```
Comment (nhjschulz) [inline lib/ZumoOta/src/App.cpp:115]: judt state the issue, not the effect. App::start is void return so I doubt the app notices.  Maybe add an App::halt() function to call after a fatal that stops in endless loop.

File: lib/ZumoOta/src/App.cpp [hunk 3]
```diff
@@ -74,45 +83,66 @@ App::~App()
...
-  WiFi.begin(ssid, password);
-    if (WiFi.waitForConnectResult() != WL_CONNECTED) {
-        Serial.printf("WiFi Failed!\n");
-        return;
-    }
-
-    Serial.print("IP Address: ");
-    Serial.println(WiFi.localIP());
-
+    Serial.begin(SERIAL_BAUDRATE);
```
Comment (nhjschulz) [inline lib/ZumoOta/src/App.cpp:124]: shouln't this be part of logInit() or do you use Serial for anything else then logging ?

File: lib/ZumoOta/src/FileManager.cpp [hunk 4]
```diff
@@ -55,7 +58,8 @@
 /******************************************************************************
  * Local Variables
  *****************************************************************************/
-
+/** Serial log sink */
+static LogSinkPrinter gLogSinkSerial("Serial", &Serial);
```
Comment (nhjschulz) [inline lib/ZumoOta/src/FileManager.cpp:62]: I doubt you need this here again as logging is a singleton. Only the one in App should be enough.

File: lib/ZumoOta/src/FileManager.cpp [hunk 5]
```diff
@@ -71,9 +75,18 @@ FileManager::~FileManager()
    
 bool FileManager::init()
 {
+    if (true == Logging::getInstance().registerSink(&gLogSinkSerial))
```
Comment (nhjschulz) [inline lib/ZumoOta/src/FileManager.cpp:78]: see former comment.

File: lib/ZumoOta/src/Upload.cpp [hunk 6]
```diff
@@ -59,14 +60,93 @@
...
+    if (final)
+    {
+        file.close();
+        request->redirect("/filelist");
+    }
+
+    // Überprüfe, ob die Datei im Dateisystem liegt
+    if (isFileUploaded(request, updatedFilename))
+    {
+        LOG_DEBUG("Die hochgeladene Datei liegt im Dateisystem.");
```
Comment (nhjschulz) [inline lib/ZumoOta/src/Upload.cpp:125]: english language shall be used.

File: lib/ZumoOta/src/Upload.cpp [hunk 7]
```diff
@@ -59,14 +60,93 @@
...
+    }
+    else
+    {
+        LOG_DEBUG("Die hochgeladene Datei liegt nicht im Dateisystem.");
```
Comment (nhjschulz) [inline lib/ZumoOta/src/Upload.cpp:129]: english

File: lib/ZumoOta/src/WebServerCustom.cpp [hunk 8]
```diff
@@ -69,9 +70,49 @@ WebServerCustom::~WebServerCustom()
 {
 }
 
+String WebServerCustom::listFiles(bool ishtml)
```
Comment (nhjschulz) [inline lib/ZumoOta/src/WebServerCustom.cpp:73]: indentation levels wrong, 2 instead of 4 spaces

File: lib/ZumoOta/src/WebServerCustom.cpp [hunk 9]
```diff
@@ -69,9 +70,49 @@ WebServerCustom::~WebServerCustom()
...
+  }
+  if (ishtml) {
+    returnText += "</table>";
+  }
+  root.close();
+  foundfile.close();
+  return returnText;
+}
+
+String WebServerCustom:: humanReadableSize(const size_t bytes)
```
Comment (nhjschulz) [inline lib/ZumoOta/src/WebServerCustom.cpp:98]: the flash size of the robot is I belive 32kb only, I strongly doubt to ever see hex files  beyond 100kb.

File: lib/ZumoOta/src/WebServerCustom.cpp [hunk 10]
```diff
@@ -69,9 +70,49 @@ WebServerCustom::~WebServerCustom()
...
 void WebServerCustom::init()
 {
-    
+   server.on("/filelist", HTTP_GET, [this](AsyncWebServerRequest *request)
+    {
+        String fileList = listFiles(true);
+        request->send(200,"text/html", fileList);
+    });
+
+    // ... Weitere Routen und Konfigurationen ...
```
Comment (nhjschulz) [inline lib/ZumoOta/src/WebServerCustom.cpp:114]: stick to english please

File: lib/ZumoOta/src/WebServerCustom.cpp [hunk 11]
```diff
@@ -89,10 +130,10 @@ void WebServerCustom::init()
 
 void WebServerCustom::handleUploadRequest()
 {
-    server.on("/upload", HTTP_POST, [](AsyncWebServerRequest *request)
-    {
-        upload.handleUploadButtonPress();
-        request->send(200, "text/plain", "Upload Button gedrueckt");
+    server.on("/upload", HTTP_POST, [this](AsyncWebServerRequest *request) {
+    }, [this](AsyncWebServerRequest *request, String filename, size_t index, uint8_t *data, size_t len, bool final) {
+        // Verarbeite den Dateiupload
```
Comment (nhjschulz) [inline lib/ZumoOta/src/WebServerCustom.cpp:135]: another german easter egg ...

**Review by nhjschulz — CHANGES_REQUESTED** [review]

File: lib/ZumoOta/src/WebServerCustom.cpp (hunk 11 above)
Comment (chezlyne) [reply lib/ZumoOta/src/WebServerCustom.cpp:135]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/WebServerCustom.cpp (hunk 10 above)
Comment (chezlyne) [reply lib/ZumoOta/src/WebServerCustom.cpp:114]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/WebServerCustom.cpp (hunk 9 above)
Comment (chezlyne) [reply lib/ZumoOta/src/WebServerCustom.cpp:98]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/WebServerCustom.cpp (hunk 8 above)
Comment (chezlyne) [reply lib/ZumoOta/src/WebServerCustom.cpp:73]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/Upload.cpp (hunk 7 above)
Comment (chezlyne) [reply lib/ZumoOta/src/Upload.cpp:129]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/FileManager.cpp (hunk 5 above)
Comment (chezlyne) [reply lib/ZumoOta/src/FileManager.cpp:78]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/FileManager.cpp (hunk 4 above)
Comment (chezlyne) [reply lib/ZumoOta/src/FileManager.cpp:62]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/App.cpp (hunk 3 above)
Comment (chezlyne) [reply lib/ZumoOta/src/App.cpp:124]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/App.cpp (hunk 2 above)
Comment (chezlyne) [reply lib/ZumoOta/src/App.cpp:115]: It will be taken care of by the next commit.

File: lib/ZumoOta/src/App.cpp (hunk 1 above)
Comment (chezlyne) [reply lib/ZumoOta/src/App.cpp:100]: It will be taken care of by the next commit.

**Comment by BlueAndi** [comment]
Please consider the static test result of the CI.

File: lib/ZumoOta/src/Upload.cpp [hunk 12]
```diff
@@ -59,14 +60,68 @@
...
+        else
+        {
+            LOG_DEBUG(String(updatedFilename) + "is not found in FileSystem.");
+        }
+            request->redirect("/filelist");
+    
+        }
+    else
+    {
+        LOG_ERROR("Please keep trying this is not the last datablock!");
```
Comment (nhjschulz) [inline lib/ZumoOta/src/Upload.cpp:115]: pointless message: who should try what ?

**Review by nhjschulz — APPROVED** [review]
Ok to check -in.

**Pull request merged**